src/Interpreters/DDLWorker.cpp

```diff
@@ -144,7 +144,14 @@ void DDLWorker::shutdown()
 
 DDLWorker::~DDLWorker()
 {
-    DDLWorker::shutdown();
+    try
+    {
+        DDLWorker::shutdown();
+    }
+    catch (...)
```

Comment:
`join` [can](https://en.cppreference.com/w/cpp/thread/thread/join) throw exception. anyway we have a report:
```
  libc.so.6           0x7f4603070a7c pthread_kill
  libc.so.6           0x7f460301c476 gsignal
  libc.so.6           0x7f46030027f3 abort
  clickhouse          0x000012c1d530 DB::DDLWorker::~DDLWorker
  clickhouse          0x00001297ef26 null
  clickhouse          0x0000129670eb DB::DatabaseReplicated::shutdown
  clickhouse          0x000012c36bc9 DB::DatabaseCatalog::shutdownImpl
  clickhouse          0x000012bbd8e4 null
  clickhouse          0x000012bbd72e DB::Context::shutdown
  clickhouse          0x00000e163261 null
  clickhouse          0x00000e158070 DB::Server::main
  clickhouse          0x00001782ea22 Poco::Util::Application::run
  clickhouse          0x00000e143439 DB::Server::run
  clickhouse          0x00001784227a Poco::Util::ServerApplication::run
  clickhouse          0x00000e140006 mainEntryClickHouseServer
  clickhouse          0x000008932f93 main
  libc.so.6           0x7f4603003d90 null
  libc.so.6           0x7f4603003e40 __libc_start_main
  clickhouse          0x000007f8a56e _start
```

Reply:
Looks like the root cause of the issue is https://github.com/ClickHouse/ClickHouse/issues/45431. Would you like to fix it, btw?

Reply:
We already have such flag, and we check `ThreadFromGlobalPoolImpl::joinable()` when required. The issue is that checks for `joinable()` are not atomic with `join()`. It can be fixed by introducing something like `joinIfJoinable`

Reply:
I meant smth like this:

``` cpp
/// Send thread will exit after sending close request or on expired flag
if (send_thread.joinable() && compare_exchange(send_joined, false, true))
    send_thread.join();
```

so we won't attempt to join this thread more than once inside this class. but there is another problem as I understood when we concurrently call `joinable` (on t0) and `join` (on t1). it might happen that t0 successfully checked that `state is true` then t1 calls `state.reset()` and then t0 calls `state->thread_id`.

so maybe it should be

``` cpp
/// Send thread will exit after sending close request or on expired flag
if (compare_exchange(send_joined, false, true) && send_thread.joinable())
    send_thread.join();
```

Reply:
Yes, it may crash when two thread concurrently do something like
```
if (thread.joinable())
    thread.join();
```
And it's fundamentally broken due to the TOCTOU issue. It can be solved with `joinIfJoinable` method that will do it atomically (with a mutex or using compare_exchange - it doesn't really matter)

Reply:
my concern with `joinIfJoinable` is it means we will implement support for an anti-pattern (because join a thread twice wouldn't work for standard threads as well). wdyt?

Reply:
Sorry, I didn't get your point. We already have a lot of places like
```
if (thread.joinable())
    thread.join();
```
and the problem is that it's not atomic. We can simply replace it with a method that do it atomically.

> standard threads

Do you mean `std::thread`? We don't use raw `std::thread`s

Reply:
my point is this code
``` cpp
if (thread.joinable())
    thread.join();
```
is ok as far as it not exec concurrently. and apparently all other places in code do not require thread-safety. `ThreadFromGlobalPoolImpl` is not thread-safe and it looks strange to change it because of one place in code (and add only one "thread-safe" method is also strange imo).

Reply:
The problem is that we already have some places where code like this can be executed concurrently. So we have two ways to fix it:
 - Avoid concurrent execution (call `~ZooKeeper()` and `DDLWorker::shutdown()` from only one thread)
 - Make it thread-safe (add a mutex or an atomic flag and compare_exchange)

I like your solution with an atomic flag and compare_exchange, but I don't understand what's the problem with putting it into a special method. It will avoid duplicating the code and will avoid new issues like this one.